messenger.rst

```diff
@@ -436,63 +548,239 @@ while still having them passed to their respective handler:
+
+.. code-block:: terminal
+
+    $ sudo supervisorctl reread
+
+    $ sudo supervisorctl update
+
+    $ sudo supervisorctl start messenger-consume
+
+See the `Supervisor docs`_ for more details.
+
+Retries & Failures
 ------------------
 
-Once your messages have been routed, you will like to consume your messages in
-most of the cases. To do so, use the ``messenger:consume`` command like this:
+If an exception is thrown while consuming a message from a transport it will
+automatically be re-sent to the transport to be tried again. By default, a message
+will be retried 3 times before being discarded or
+:ref:`sent to the failure transport <messenger-failure-transport>`. Each retry
+will also be delayed, in case the failure was due to a temporary issue. All of
+this is configurable for each transport:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/messenger.yaml
+        framework:
+            messenger:
+                transports:
+                    async_priority_high:
+                        dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
+
+                        # default configuration
+                        retry_strategy:
+                            max_retries: 3
+                            # milliseconds delay
+                            delay: 1000
+                            # causes the delay to be higher before each retry
+                            # e.g. 1 second delay, 2 seconds, 4 seconds
+                            multiplier: 2
+                            max_delay: 0
+                            # override all of this with a service that
+                            # implements Symfony\Component\Messenger\Retry\RetryStrategyInterface
+                            # service: null
+
+Avoiding Retrying
+~~~~~~~~~~~~~~~~~
+
+Sometimes handling a message might fail in a way that you *know* is permanent
+and should not be retried. If you throw
+:class:`Symfony\\Component\\Messenger\\Exception\\UnrecoverableMessageHandlingException`,
+the message will not be retried.
+
+.. _messenger-failure-transport:
+
+Saving & Retrying Failed Messages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a message fails it is retried multiple times (``max_retries``) and then will
+be discarded. To avoid this happening, you can instead configure a ``failure_transport``:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/messenger.yaml
+        framework:
+            messenger:
+                # after retrying, messages will be sent to the "failed" transport
+                failure_transport: failed
+
+                transports:
+                    # ... other transports
+
+                    failed: 'doctrine://default?queue_name=failed'
+
+In this example, if handling a message fails 3 times (default ``max_retries``),
+it will then be sent to the ``failed`` transport. While you *can* use
+``messenger:consume failed`` to consume this like a normal transport, you'll
```

Comment:
It would be good to mention here that failed messages that fail again for the configured retry amount, do not go back into the failed transport and will be discarded.